### PR TITLE
fix(clients): reject outstanding requests to dropped clients

### LINKED ISSUE #9462: Reject outstanding requests to dropped clients
If a request is sent to a client and that client is dropped (due to being disconnected for too long), the request will eventually timeout rather than being promptly rejected at the time the client is dropped.

Ideally, we would reject these outstanding requests at the time the client is dropped.

We could implement this by tracking outstanding requests in the Gateway's `ClientData`, removing them when the client sends a response, and rejecting them when the client is dropped.

### PR DESCRIPTION
**Pull request by ReubenBond**
Fixes #9462
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/orleans/pull/9463)

### CODE REVIEW

File: src/Orleans.Runtime/Messaging/Gateway.cs [hunk 1]
```diff
@@ -187,6 +187,11 @@ internal SiloAddress TryToReroute(Message msg)
 
             if (msg.Direction != Message.Directions.Response)
             {
+                if (ClientGrainId.TryParse(msg.SendingGrain, out var clientId) && clients.TryGetValue(clientId, out var clientState))
+                {
+                    clientState.OnClientResponse(msg);
```
Comment (benjaminpetit) [inline src/Orleans.Runtime/Messaging/Gateway.cs:192]: Isn't it possible that the client is still connected to other silos in the cluster? Or is it an edge case?

File: test/Orleans.Runtime.Tests/ClientConnectionTests/ClientDisconnectionTests.cs [hunk 2]
```diff
@@ -0,0 +1,219 @@
+#nullable enable
+using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Orleans.GrainReferences;
+using Orleans.Runtime;
+using Orleans.TestingHost;
+using Orleans.TestingHost.Utils; // For TestClusterPortAllocator
+using TestGrainInterfaces; // Assuming this is accessible from Tester project
+using UnitTests.GrainInterfaces;
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Orleans.Runtime.Tests/ClientConnectionTests/ClientDisconnectionTests.cs]: Removed the unused imports in e536961aa.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.TestingHost/InProcTestCluster.cs [hunk 3]
```diff
@@ -356,38 +267,19 @@ public IEnumerable<InProcessSiloHandle> GetActiveSilos()
...
-        {
-            WriteLog("WaitForLivenessToStabilize observed stable active silo and gateway views");
-        }
-        else
-        {
-            WriteLog("WaitForLivenessToStabilize reached the fallback wait of {0}", stabilizationTime);
-        }
+        WriteLog(Environment.NewLine + Environment.NewLine + "WaitForLivenessToStabilize is about to sleep for {0}", stabilizationTime);
+        await Task.Delay(stabilizationTime);
+        WriteLog("WaitForLivenessToStabilize is done sleeping");
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.TestingHost/InProcTestCluster.cs]: Restored active-silo, gateway, and grain-directory convergence checks in e536961aa.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.TestingHost/InProcTestCluster.cs [hunk 4]
```diff
@@ -413,7 +324,7 @@ public static TimeSpan GetLivenessStabilizationTime(ClusterMembershipOptions clu
         if (didKill)
         {
             // in case of hard kill (kill and not Stop), we should give silos time to detect failures first.
-            stabilizationTime = TestingUtils.Multiply(clusterMembershipOptions.MaxProbeTimeout, clusterMembershipOptions.NumMissedProbesLimit);
+            stabilizationTime = TestingUtils.Multiply(clusterMembershipOptions.ProbeTimeout, clusterMembershipOptions.NumMissedProbesLimit);
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.TestingHost/InProcTestCluster.cs]: Hard-kill stabilization now uses `MaxProbeTimeout` in e536961aa.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/Messaging/Gateway.cs [hunk 5]
```diff
@@ -200,6 +200,11 @@ internal void RecordClosedConnection(GatewayInboundConnection connection)
...
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Messaging/Gateway.cs]: Moved response tracking ahead of client-to-client routing so responses from clients to server targets clear outstanding requests in e536961aa.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.TestingHost/InProcTestCluster.cs [hunk 6]
```diff
@@ -700,18 +618,49 @@ public async Task InitializeClientAsync()
...
         }
-        catch
+
+        return null;
+    }
+
+    /// <summary>
+    /// Gets the client with the given name, creating a new client if none exists with that name.
+    /// </summary>
+    public async Task<IClusterClient> GetClientAsync(string name, Action<IHostApplicationBuilder>? configure = null)
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.TestingHost/InProcTestCluster.cs:668]: Applied the optional client-host configurator when creating named clients in e536961aa.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.TestingHost/InProcTestCluster.cs [hunk 7]
```diff
@@ -963,9 +903,14 @@ public void Dispose()
         }
 
         ClientHost?.Dispose();
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.TestingHost/InProcTestCluster.cs]: Removed the duplicate default-client disposal path in e536961aa.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.TestingHost/InProcTestCluster.cs [hunk 8]
```diff
@@ -781,24 +730,11 @@ public async Task<InProcessSiloHandle> CreateSiloAsync(InProcessTestSiloSpecific
...
-                    }
-                }
-
-#pragma warning disable ORLEANSEXP003
-                if (Options.UseDistributedGrainDirectory)
-#pragma warning restore ORLEANSEXP003
-                {
-                    new ConfigureDistributedGrainDirectory().Configure(siloBuilder);
+                    siloBuilder.AddGrainDirectory(GrainDirectoryAttribute.DEFAULT_GRAIN_DIRECTORY, (_, _) => _grainDirectory);
                 }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.TestingHost/InProcTestCluster.cs:784]: Restored the `UseTestClusterGrainDirectory` and `UseDistributedGrainDirectory` option semantics in e536961aa.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Core/SystemTargetInterfaces/IManagementGrain.cs [hunk 9]
```diff
@@ -141,6 +141,13 @@ public interface IManagementGrain : IGrainWithIntegerKey, IVersionManager
         /// <param name="hostsIds">The hosts to invoke the operation on.</param>
         /// <returns>A task representing the work performed.</returns>
         ValueTask ResetGrainCallFrequencies(SiloAddress[]? hostsIds = null);
+
+        /// <summary>
+        /// Instructs all gateways to drop defunct (disconnected and expired) clients.
+        /// </summary>
+        /// <param name="excludeRecent">If true, only clients that have been disconnected for longer than the configured client expiration time will be dropped.</param>
+        /// <returns>A task representing the work performed.</returns>
+        Task DropDisconnectedClients(bool excludeRecent);
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Core/SystemTargetInterfaces/IManagementGrain.cs:150]: Regenerated the Orleans.Core public API surface in e536961aa.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/Messaging/Gateway.cs [hunk 10]
```diff
@@ -229,17 +236,17 @@ private bool IsTargetingLocalGateway(SiloAddress siloAddress)
         }
 
         // There is NO need to acquire individual ClientState lock, since we only close an older socket.
-        internal void DropDisconnectedClients()
+        internal void DropDisconnectedClients(bool excludeRecent = false)
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Messaging/Gateway.cs]: Changed the default to `excludeRecent: true`, preserving maintenance-loop behavior while management calls can still explicitly drop every disconnected client.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/Messaging/Gateway.cs [hunk 11]
```diff
@@ -410,17 +426,29 @@ private async Task RunMessageLoop()
...
+                                        _outstandingRequestsToClient[(message.SendingGrain, message.Id)] = message;
+                                    }
+
+                                    LogTraceSentQueuedMessage(_gateway.logger, message, Id);
+                                }
+                                else
+                                {
+                                    _pendingToSend.Enqueue((WorkItemType.SendMessageToClient, message));
+                                    return;
+                                }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Messaging/Gateway.cs:447]: The send loop now breaks out of the current drain pass instead of returning permanently. `RecordClosedConnection` signals the loop, so it remains alive to reject queued and outstanding requests when the client is dropped or resume delivery after reconnection.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.TestingHost/InProcTestCluster.cs [hunk 12]
```diff
@@ -781,7 +746,6 @@ public async Task<InProcessSiloHandle> CreateSiloAsync(InProcessTestSiloSpecific
 
                 siloBuilder.Services
                     .Configure<HostOptions>(options => options.ShutdownTimeout = TimeSpan.FromSeconds(30));
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.TestingHost/InProcTestCluster.cs:765]: Restored both `TestClusterFatalErrorHandler.Configure` before the in-process silo is built and `Attach` immediately afterward, so fatal errors stop the affected test silo instead of terminating the test runner.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Comment by ReubenBond** [comment]
The post-review CI run exposed a PR-specific multi-client lifecycle regression: both `InProcessTestCluster_Client_RemainsUnavailableWhenStartupFails` and `InProcessTestCluster_Client_IsAvailableOnlyWhileDeployed` failed on every OS/TFM because client hosts were published before startup completed and the public unavailable-state message changed. Commit `5465c0c3d` now publishes hosts only after successful startup, disposes failed hosts, and preserves the established client availability contract. Run: https://github.com/dotnet/orleans/actions/runs/32148333736

File: src/Orleans.Runtime/Core/InsideRuntimeClient.cs [hunk 13]
```diff
@@ -645,6 +645,16 @@ private async Task MonitorCallbackExpiry()
...
 
+        public void Dispose()
+        {
+            foreach (var callback in callbacks)
+            {
+                var message = callback.Value.Message;
+                var response = messageFactory.CreateRejectionResponse(message, Message.RejectionTypes.Unrecoverable, "Host is shutting down.", null);
+                callback.Value.OnResponse(response);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Core/InsideRuntimeClient.cs:673]: `Dispose()` now uses the existing shutdown sweep. Each callback completes through `OnHostShutdown`, which unregisters it from the dictionary before resolving the caller, preventing retained or repeated completions.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/Networking/GatewayInboundConnection.cs [hunk 14]
```diff
@@ -170,7 +170,7 @@ public void FailMessage(Message msg, string reason)
                     msg,
                     Message.RejectionTypes.Transient,
                     $"Silo {this.myAddress} is rejecting message: {msg}. Reason = {reason}",
-                    new SiloUnavailableException());
+                    new ClientNotAvailableException(reason ?? "Connection terminated."));
             }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Networking/GatewayInboundConnection.cs:187]: The rejection now identifies `msg.TargetGrain` in `ClientNotAvailableException` before appending the transport reason, so failures name the affected client target.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.TestingHost/InProcTestCluster.cs [hunk 15]
```diff
@@ -60,32 +63,49 @@ public ReadOnlyCollection<InProcessSiloHandle> Silos
...
-    /// <exception cref="InvalidOperationException">The cluster has not been deployed or the client has been stopped.</exception>
-    public IClusterClient Client => InternalClient ?? throw new InvalidOperationException(
-        "The test cluster client is unavailable because the cluster has not been deployed or has been stopped.");
+    public IClusterClient Client => InternalClient;
 
     /// <summary>
     /// The port allocator.
     /// </summary>
-    public ITestClusterPortAllocator PortAllocator { get; }
+    public ITestClusterPortAllocator? PortAllocator { get; }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.TestingHost/InProcTestCluster.cs]: Made `PortAllocator` non-nullable and replaced conditional disposal with direct disposal, matching the constructor invariant.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.TestingHost/InProcTestSiloSpecificOptions.cs [hunk 16]
```diff
@@ -40,7 +40,7 @@ public static InProcessTestSiloSpecificOptions Create(InProcessTestCluster testC
 
         if (assignNewPort)
         {
-            var (siloPort, gatewayPort) = testCluster.PortAllocator.AllocateConsecutivePortPairs(1);
+            var (siloPort, gatewayPort) = testCluster.PortAllocator!.AllocateConsecutivePortPairs(1);
             result.SiloPort = siloPort;
             result.GatewayPort = (instanceNumber == 0 || testClusterOptions.GatewayPerSilo) ? gatewayPort : 0;
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.TestingHost/InProcTestSiloSpecificOptions.cs:45]: Removed the unnecessary null-forgiving operator. `PortAllocator` is used directly under its non-nullable contract.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.TestingHost/InProcTestCluster.cs [hunk 17]
```diff
@@ -939,10 +1029,13 @@ await Task.Run(async () =>
                 await DisposeAsync(handle).ConfigureAwait(false);
             }
 
-            await DisposeAsync(ClientHost).ConfigureAwait(false);
-            ClientHost = null;
+            foreach (var clientHost in _clientHosts.Values)
+            {
+                await DisposeAsync(clientHost).ConfigureAwait(false);
+            }
+            _clientHosts.Clear();
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.TestingHost/InProcTestCluster.cs]: Updated disposal to use one shared completion task and serialize registry capture with client lifecycle operations. New client operations are rejected once disposal starts, concurrent disposers await the same teardown, and cleanup attempts every silo, client, allocator, and observer before aggregating any failures.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: test/Orleans.Runtime.Tests/ClientConnectionTests/ClientDisconnectionTests.cs [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+
+        // B disconnects
+        await _cluster.RemoveClientAsync("ClientB");
+
+        // B's pending request should be promptly rejected locally.
+        var exception = await Assert.ThrowsAnyAsync<Exception>(() => responseTask);
+        Assert.True(
+            exception is OperationCanceledException or OrleansMessageRejectionException,
+            $"Unexpected exception type: {exception.GetType()}");
+
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Orleans.Runtime.Tests/ClientConnectionTests/ClientDisconnectionTests.cs:181]: Included `SiloUnavailableException` in the accepted shutdown outcomes, covering the `CallbackData.OnHostShutdown()` completion path while retaining the existing cancellation and rejection paths.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/Networking/GatewayConnectionListener.cs [hunk 19]
```diff
@@ -44,7 +43,6 @@ public GatewayConnectionListener(
             this.messageCenter = messageCenter;
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Networking/GatewayConnectionListener.cs:41]: Removed the unused logger constructor parameter and its using directive. The listener continues to resolve through dependency injection with the remaining constructor dependencies.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Core/Messaging/Message.cs [hunk 20]
```diff
@@ -257,6 +271,169 @@ public Dictionary<string, object>? RequestContextData
...
+                context.Remove(GatewayResponseTargetHeader);
+                if (context.Count == 0)
+                {
+                    RequestContextData = null;
+                }
+            }
+
+            _gatewayRequestSource = null;
+            _hasGatewayRequestSource = false;
+        }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Core/Messaging/Message.cs:322]: Split full message reset from client-ingress routing sanitization. `ClearGatewayRequestOwner` now also clears the gateway timeout and trusted-target flag; ingress uses `ClearGatewayRequestRouting` so the trusted caller timeout remains available for tracking.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/Networking/GatewayInboundConnection.cs [hunk 21]
```diff
@@ -164,13 +196,15 @@ public void FailMessage(Message msg, string reason)
...
+                this.gateway.RemoveTrackedClientRequest(msg);
+                msg.RestoreGatewayRequestSource();
 
                 // Done retrying, send back an error instead
                 this.messageCenter.SendRejection(
                     msg,
                     Message.RejectionTypes.Transient,
                     $"Silo {this.myAddress} is rejecting message: {msg}. Reason = {reason}",
-                    new SiloUnavailableException());
+                    new ClientNotAvailableException($"Target client {msg.TargetGrain} is unavailable. {reason ?? "Connection terminated."}"));
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Networking/GatewayInboundConnection.cs]: Updated the rejection details to identify the unavailable target client and retain the message/reason context, matching the `ClientNotAvailableException` semantics.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/Core/InsideRuntimeClient.cs [hunk 22]
```diff
@@ -207,16 +210,27 @@ public void SendRequest(
...
-            // calls. Reject those calls so that they cannot outlive the shutdown callback sweep.
-            if (Volatile.Read(ref _isStopping) != 0)
+            var rejectForShutdown = false;
+            lock (_requestAdmissionLock)
             {
-                callbackData?.OnHostShutdown();
-                return;
+                // Completing callbacks during shutdown can resume application code which issues follow-up
+                // calls. Admission and sending are serialized with the stopping transition so that no
+                // request, including a one-way request, can be sent after shutdown begins.
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Core/InsideRuntimeClient.cs]: Replaced the global monitor with a `ReaderWriterLockSlim`: normal sends take concurrent read admission, while lifecycle/fallback shutdown takes the exclusive write side before setting the stopping invariant and disposing the timer. This preserves the race guarantee without serializing the hot path.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/Core/InsideRuntimeClient.cs [hunk 23]
```diff
@@ -176,9 +177,11 @@ public void SendRequest(
                 sharedData = this.sharedCallbackData;
             }
 
+            var responseTimeout = request.GetDefaultResponseTimeout() ?? sharedData.ResponseTimeout;
+            message.SetGatewayRequestTimeout(responseTimeout);
             if (this.messagingOptions.DropExpiredMessages && message.IsExpirableMessage())
             {
-                message.TimeToLive = request.GetDefaultResponseTimeout() ?? sharedData.ResponseTimeout;
+                message.TimeToLive = responseTimeout;
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Core/InsideRuntimeClient.cs:191]: Changed the silo runtime path in a3069c2dd7 to add gateway timeout metadata only for client/observer targets, avoiding request-context allocation for ordinary grain calls.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Core/Runtime/OutsideRuntimeClient.cs [hunk 24]
```diff
@@ -281,11 +281,12 @@ public void SendRequest(GrainReference target, IInvokable request, IResponseComp
                 message.TargetSilo = systemTargetGrainId.GetSiloAddress();
             }
 
+            var responseTimeout = request.GetDefaultResponseTimeout() ?? this.sharedCallbackData.ResponseTimeout;
+            message.SetGatewayRequestTimeout(responseTimeout);
             if (this.clientMessagingOptions.DropExpiredMessages && message.IsExpirableMessage())
             {
                 // don't set expiration for system target messages.
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Core/Runtime/OutsideRuntimeClient.cs:292]: Applied the same client-target guard to the outside runtime in a3069c2dd7, so regular external grain calls no longer carry the gateway timeout header.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Core/Runtime/RequestContextExtensions.cs [hunk 25]
```diff
@@ -20,7 +20,9 @@ public static void Import(Dictionary<string, object>? contextData)
         {
             var values = contextData switch
             {
-                { Count: > 0 } => contextData.ToDictionary(kvp => kvp.Key, kvp => kvp.Value),
+                { Count: > 0 } => contextData
+                    .Where(static kvp => !Message.IsGatewayRequestContextHeader(kvp.Key))
+                    .ToDictionary(kvp => kvp.Key, kvp => kvp.Value),
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Core/Runtime/RequestContextExtensions.cs]: Fixed in cfa817e913. Gateway headers are filtered with a single pass and the destination dictionary is allocated only when a non-gateway entry is encountered, including the all-reserved-header case.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/Messaging/Gateway.cs [hunk 26]
```diff
@@ -400,27 +499,38 @@ private async Task RunMessageLoop()
...
 
-                        var connection = Volatile.Read(ref _connection);
-                        if (connection is null)
+                        var pendingCount = _pendingToSend.Count;
+                        while (pendingCount-- > 0 && _pendingToSend.TryDequeue(out var message))
                         {
-                            continue;
-                        }
+                            var connection = Volatile.Read(ref _connection);
+                            if (connection is null)
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Messaging/Gateway.cs]: Fixed in cfa817e913. The gateway loop now checks the connection once before draining, uses `TryDequeue` until empty, and stops immediately after a failed send requeues the message, avoiding `ConcurrentQueue.Count` and dequeue/re-enqueue churn while disconnected.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/Core/InsideRuntimeClient.cs [hunk 27]
```diff
@@ -645,6 +669,26 @@ private async Task MonitorCallbackExpiry()
...
+            try
+            {
+                Volatile.Write(ref _isStopping, 1);
+                this.callbackTimer.Dispose();
+            }
+            finally
+            {
+                _requestAdmissionLock.ExitWriteLock();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Core/InsideRuntimeClient.cs:706]: Replaced the owned `ReaderWriterLockSlim` with a non-allocating atomic admission protocol. The sign bit closes admission and the remaining bits track active sends, so shutdown prevents new sends and waits for in-flight sends without an owned disposable lock or a global serialized hot path.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Runtime/Networking/GatewayInboundConnection.cs [hunk 28]
```diff
@@ -164,13 +175,15 @@ public void FailMessage(Message msg, string reason)
...
+                msg.RestoreGatewayRequestSource();
 
                 // Done retrying, send back an error instead
                 this.messageCenter.SendRejection(
                     msg,
                     Message.RejectionTypes.Transient,
-                    $"Silo {this.myAddress} is rejecting message: {msg}. Reason = {reason}",
-                    new SiloUnavailableException());
+                    $"Target client {msg.TargetGrain} is unavailable. Message: {msg}. Reason = {reason}",
+                    new ClientNotAvailableException($"Target client {msg.TargetGrain} is unavailable. {reason ?? "Connection terminated."}"));
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/Networking/GatewayInboundConnection.cs]: Removed the redundant null-coalescing fallback and now use the non-null `reason` directly in `ClientNotAvailableException`.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).
